File: dotnet/src/Messaging.Scenarios/S3Cancel.cs
using Messaging.Abstractions;
using static Messaging.Scenarios.ScenarioHelpers;

namespace Messaging.Scenarios;

/// <summary>
/// S3 — the key discriminator from the research: cancel a scheduled message
/// before it fires and prove it never arrives. Adapters that cannot do this
/// surface an <see cref="OperationNotSupportedException"/>, recorded as an honest
/// ⊘ rather than ✗.
/// </summary>
public sealed class CancelScheduled : IScenario
{
    private static readonly TimeSpan Delay = TimeSpan.FromMilliseconds(3000);

    public string Name => "S3 cancel scheduled";
    public string Description => "A scheduled message is cancelled and never delivered.";

    public async Task<ScenarioResult> RunAsync(IMessageScheduler s)
    {
        var t0 = StartClock();
        var dest = $"mbc.s3.{Nonce()}";
        var token = $"cancel-{Nonce()}";
        await using var collector = await MessageCollector.StartAsync(s, dest);

        var handle = await s.ScheduleAsync(dest, token, DateTimeOffset.UtcNow + Delay);

        try
        {
            await s.CancelAsync(handle);
        }
        catch (OperationNotSupportedException)
        {
            return Unsupported(Name, "no cancel API on this broker", t0);
        }

        // Wait past the original delivery time; the message must never appear.
        await Task.Delay(Delay + TimeSpan.FromSeconds(2));

        return collector.Bodies().Contains(token)
            ? Fail(Name, "cancel returned but message still delivered", t0)
            : Pass(Name, "scheduled message successfully cancelled", t0);
    }
}

File: typescript/src/scenarios/s3-cancel.ts
import { IMessageScheduler, NotSupportedError } from '../abstractions';
import {
  MessageCollector,
  Scenario,
  ScenarioResult,
  delay,
  fail,
  nonce,
  pass,
  unsupported,
} from './scenario';

const DELAY_MS = 3000;

/** S3 — the key discriminator from the research: cancel a scheduled message
 *  before it fires and prove it never arrives. Adapters that cannot do this
 *  surface a NotSupportedError, which we record as an honest ⊘ rather than ✗. */
export const cancelScheduled: Scenario = {
  name: 'S3 cancel scheduled',
  description: 'A scheduled message is cancelled and never delivered.',
  async run(s: IMessageScheduler): Promise<ScenarioResult> {
    const t0 = performance.now();
    const dest = `mbc.s3.${nonce()}`;
    const token = `cancel-${nonce()}`;
    const collector = await MessageCollector.start(s, dest);
    try {
      const handle = await s.schedule(
        dest,
        token,
        new Date(Date.now() + DELAY_MS),
      );

      try {
        await s.cancel(handle);
      } catch (err) {
        if (err instanceof NotSupportedError) {
          return unsupported(this.name, 'no cancel API on this broker', t0);
        }
        throw err;
      }

      // Wait past the original delivery time; the message must never appear.
      await delay(DELAY_MS + 2000);
      return collector.bodies().includes(token)
        ? fail(this.name, 'cancel returned but message still delivered', t0)
        : pass(this.name, 'scheduled message successfully cancelled', t0);
    } finally {
      await collector.stop();
    }
  },
};
